storage/s3.go

```diff
@@ -32,6 +32,11 @@ var _ Storage = (*S3)(nil)
 
 var sentinelURL = urlpkg.URL{}
 
+var (
+	mutex            = &sync.Mutex{}
+	mapS3optsSession = map[S3Options]*session.Session{}
```

Comment:
Instead of keeping global mutex and map, we can create a new type like:
```go
type Session struct{
      sync.RWMutex
      sessions map[S3Options]*session.Session{}
}

func (s Session) Read(opts S3Options) *session.Session {
     s.RLock()
     defer s.RUnlock()
     return s.sessions[opts]
}

func (s *Session) Write(opts S3Options) error {
      s.Lock()
      defer s.Unlock()

      awsSession, err := newSession(opts)
      if err != nil {
	  return err
      }

      s.sessions[opts] = awsSession
      return nil
}

func (s *Session) Clear() {
        s.Lock()
	defer s.Unlock()
	s.sessions = map[S3Options]*session.Session{}
}
```